### PR TITLE
[SPARK-35908][SQL] Remove repartition if the child maximum number of rows less than or equal to 1

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

Enhance `OptimizeRepartition` to remove repartition if the child maximum number of rows less than or equal to 1. For example:
```scala
spark.sql("select count(*) from range(1, 10, 2, 2)").repartition(10).explain("cost")
```

Before this PR:
```
== Optimized Logical Plan ==
Repartition 10, true, Statistics(sizeInBytes=16.0 B, rowCount=1)
+- Aggregate [count(1) AS count(1)#2L], Statistics(sizeInBytes=16.0 B, rowCount=1)
   +- Project, Statistics(sizeInBytes=20.0 B)
      +- Range (1, 10, step=2, splits=Some(2)), Statistics(sizeInBytes=40.0 B, rowCount=5)
```

After this PR:
```
== Optimized Logical Plan ==
Aggregate [count(1) AS count(1)#2L], Statistics(sizeInBytes=16.0 B, rowCount=1)
+- Project, Statistics(sizeInBytes=20.0 B)
   +- Range (1, 10, step=2, splits=Some(2)), Statistics(sizeInBytes=40.0 B, rowCount=5)
```

### Why are the changes needed?

Improve query performance.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -24,7 +24,7 @@ import org.apache.spark.sql.catalyst.catalog.{InMemoryCatalog, SessionCatalog}
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.expressions.aggregate._
 import org.apache.spark.sql.catalyst.plans._
-import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.plans.logical.{RepartitionOperation, _}
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:27]: Why do we need this change here?

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Removed it.

**Review by maropu — APPROVED** [review]
cc: @HyukjinKwon

**Comment by HyukjinKwon** [comment]
I guess it's fine ... but I would be great to have a second look from people like @cloud-fan or @maryannxue

**Comment by cloud-fan** [comment]
I'm not very sure about it. One row doesn't mean one partition, right? The dataset can have many empty partitions.

**Comment by HyukjinKwon** [comment]
yeah I have the exactly same concern on the empty partitions. I can't currently think that would matter since we won't execute anything or write out something. Maybe Hadoop APIs could be affected since we didn't fix it about empty files being skipped IIRC.

I remember we have faced several corner cases related to the empty files, and the conclusion is that we're fine as long as it writes at least one empty file? (e.g., to keep the metadata like schema in parquet footer)

**Comment by wangyum** [comment]
Thank you all. Close it.

**Pull request closed without merging**